Fullscreen API implementation

theIDinside opened the pull request:
This PR implements the Fullscreen API. 

The final commit also adds some Qt UI (an "exit fullscreen" button that animates from the top down, sort of like how Chrome does it, and also Firefox). For Qt-backends, the escape key also exits out of fullscreen, fully.

The spec can be found [here](https://fullscreen.spec.whatwg.org/).

New web platform test results (fullscreen/api) with this patch series applied:

```
Ran 57 tests finished in 13.7 seconds.
  • 36 ran as expected. 0 tests skipped.
  • 1 tests had errors unexpectedly
  • 1 tests timed out unexpectedly
  • 21 tests had unexpected subtest results
```
(up from 0 expected to 36. Another test succeeds when #4329 is applied)

Additional work that needs to happen:
- https://fullscreen.spec.whatwg.org/#dom-document-fullscreenenabled needs to be implemented fully, where a document's "allowed to use" can be deterimined by the `allowFullscreen` attribute set on an iframe. This attribute can not be changed dynamically and once set for a document, is set for the rest of it's life time.

Additional work that's required can be determined by the WPT suite.

ADKaster commented:
First comment before review: Make sure your commit messages follow the guidelines. e.g. Imperative verbs, not past tense. `Do the thing` not `Did the thing`

ADKaster commented:
The missing visit on document::m_fullscreen_element needs fixed: https://github.com/LadybirdBrowser/ladybird/actions/runs/14411993971/job/40447984611?pr=4330#step:11:1711

Make sure your IDE doesn't add `""` includes. `Files that #include a missing local file: Libraries/LibWeb/CSS/SelectorEngine.cpp Libraries/LibWeb/DOM/Element.cpp UI/Qt/BrowserWindow.cpp`

And be sure to use the same version of clang-format as CI.

You should invest in `pre-commit` <https://github.com/LadybirdBrowser/ladybird/blob/master/CONTRIBUTING.md#commit-hooks>

theIDinside commented:
> The missing visit on document::m_fullscreen_element needs fixed: https://github.com/LadybirdBrowser/ladybird/actions/runs/14411993971/job/40447984611?pr=4330#step:11:1711
> 
> Make sure your IDE doesn't add `""` includes. `Files that #include a missing local file: Libraries/LibWeb/CSS/SelectorEngine.cpp Libraries/LibWeb/DOM/Element.cpp UI/Qt/BrowserWindow.cpp`
> 
> And be sure to use the same version of clang-format as CI.
> 
> You should invest in `pre-commit` https://github.com/LadybirdBrowser/ladybird/blob/master/CONTRIBUTING.md#commit-hooks

Thanks! i'll do that, I thought I had snuffed out all the style issues, but clearly I did not.

theIDinside commented:
> @theIDinside looks like the linter check is failing in idl file, it's likely an indentation issue. It should be:
> 
> ```
> enum FullscreenNavigationUI {
>     "auto",
>     "show",
>     "hide"
> };
> 
> dictionary FullscreenOptions {
>     FullscreenNavigationUI navigationUI = "auto";
> };
> ```
> 
> 4 spaces indent. Maybe in other places too

Fixed it. Rebased on most current master.

ADKaster commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -6339,6 +6340,181 @@ bool Document::fullscreen_enabled() const
+        unfullscreen_element(element);
+    }
+
+    HTML::TemporaryExecutionContext context(relevant_settings_object().realm(), HTML::TemporaryExecutionContext::CallbacksEnabled::Yes);
```
why is this necessary? the caller should take care of this.

theIDinside replied:
Will be removed (as discussed on discord).

ADKaster commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -6339,6 +6340,181 @@ bool Document::fullscreen_enabled() const
+    }
+    // 3. Let resize be false.
+    bool resize = false;
+    // 4. Let docs be the result of collecting documents to unfullscreen given doc.
```
don't be afraid of newlines. this method is way too cramped.
```
// N. aaaa
aaaaaa

// N+1. bbbb
bbbbb

// N+2 cccc
ccccc
```

ADKaster commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -6339,6 +6340,181 @@ bool Document::fullscreen_enabled() const
+    return total == 1;
+}
+
+Vector<GC::Ptr<Document>> Document::collect_documents_to_unfullscreen() const
```
spec link?

ADKaster commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -6339,6 +6340,181 @@ bool Document::fullscreen_enabled() const
+
+bool Document::is_simple_fullscreen_document() const
+{
+    u32 total = 0;
```
why are we counting here? are there spec steps involved? if not, this should just be `bool has_fullscreen_element`, with |= in the loop. `if element_is_fullscreen && has_fullscreen_element: return false` or something. adding bools to ints is icky.

theIDinside replied:
We're counting because a simple fullscreen document is one with only 1 element in it's top layer that has it's fullscreen flag set. I'll add a spec link and also just ++ increment.

ADKaster commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -6339,6 +6340,181 @@ bool Document::fullscreen_enabled() const
+Vector<GC::Ptr<Document>> Document::collect_documents_to_unfullscreen() const
+{
+    // 1. Let docs be an ordered set consisting of doc.
+    Vector<GC::Ptr<Document>> docs;
```
MarkedVector. But also "ordered set" means AK::HashTable

theIDinside replied:
I've changed it to `GC::RootVector` - yes, the algorithm states "ordered set", but the algorithm itself also maintains that order, so I see no need to make it a hash table; though I did not get a response wrt to this on discord. Closing as resolved.

ADKaster commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -6339,6 +6340,181 @@ bool Document::fullscreen_enabled() const
+{
+    // To unfullscreen an element, unset element’s fullscreen flag and iframe fullscreen flag (if any), and remove from the top layer immediately given element.
+    element->set_fullscreen_flag(false);
+    HTML::HTMLIFrameElement* iframe_element = as_if<HTML::HTMLIFrameElement>(element.ptr());
+    if (iframe_element)
+        iframe_element->set_iframe_fullscreen_flag(false);
```
```
    if (auto* iframe_element = as_if<HTML::HTMLIFrameElement>(element.ptr()))
        iframe_element->set_iframe_fullscreen_flag(false);
```

ADKaster commented on Libraries/LibWeb/DOM/Element.cpp:
```diff
@@ -2089,6 +2090,32 @@ GC::Ref<WebIDL::Promise> Element::request_fullscreen()
     return promise;
 }
 
+void Element::removing_steps_fullscreen()
```
spec link?

ADKaster commented on Libraries/LibWeb/DOM/Element.cpp:
```diff
@@ -2089,6 +2090,32 @@ GC::Ref<WebIDL::Promise> Element::request_fullscreen()
+                // 2. Otherwise, unfullscreen node.
+                doc.unfullscreen_element(*element);
+            }
+            // 3. If document’s top layer contains node, remove from the top layer immediately given node (handled above)
```
...above?

theIDinside replied:
It gets handled in both branches eventually. The spec seems to have an artefact here. Meaning, it will always be removed from the top layer.

ADKaster commented on UI/Qt/BrowserWindow.cpp:
```diff
@@ -32,12 +34,133 @@
+#include <QTimer>
+#include <QWidget>
 #include <QWindow>
+#include <qnamespace.h>
```
not sure what qnamespace is, but I doubt we want it.

ADKaster reviewed (changes requested):
I didn't look closely at the Qt changes.

As I said on discord several times, this PR and change set is much too large for a first PR. There are far too many style issues here to give the actual changes a thorough review.

Please split this up into manageable and mergable chunks while addressing these comments.

theIDinside commented on Libraries/LibWeb/DOM/Element.h:
```diff
@@ -471,6 +471,8 @@ class Element
     void release_pointer_capture(WebIDL::Long pointer_id);
     bool has_pointer_capture(WebIDL::Long pointer_id);
 
+    bool is_fullscreen_element() const;
```
Do you want me to also move the fullscreen flag to `Node`? Can there be an argument made for moving the entire algorithm over on to `Node` instead of `Element`?

theIDinside commented on Libraries/LibWeb/DOM/Element.idl:
```diff
@@ -33,6 +33,17 @@ dictionary CheckVisibilityOptions {
     boolean visibilityProperty = false;
 };
 
+// https://fullscreen.spec.whatwg.org/#api
+enum FullscreenNavigationUI {
+    "auto",
+    "show",
+    "hide"
+};
+
+dictionary FullscreenOptions {
+    FullscreenNavigationUI navigationUI = "auto";
+};
```
I'll remove this entirely probably from the PR, as it will need changes to existing IPC messages (enter fullscreen) and since complexity is already an issue with this PR, we'll leave that for the future instead.

theIDinside commented on Libraries/LibWeb/DOM/Document.h:
```diff
@@ -915,6 +915,15 @@ class Document
     [[nodiscard]] WebIDL::CallbackType* onfullscreenerror();
     void set_onfullscreenerror(WebIDL::CallbackType*);
 
+    // https://fullscreen.spec.whatwg.org/#fullscreen-an-element
+    void fullscreen_element_within_doc(GC::Ref<Element> element);
+    // https://fullscreen.spec.whatwg.org/#fullscreen-element
+    GC::Ptr<Element> fullscreen_element() const;
+    // https://fullscreen.spec.whatwg.org/#dom-document-fullscreen
+    bool fullscreen() const;
+    // https://fullscreen.spec.whatwg.org/#dom-document-fullscreenenabled
+    bool fullscreen_enabled() const;
```
What's the rule here? There are spec links in the header files, are they special cases?

theIDinside commented on Libraries/LibWeb/DOM/Element.cpp:
```diff
@@ -1905,6 +1910,185 @@ WebIDL::ExceptionOr<void> Element::insert_adjacent_html(String const& position,
+        relevant_global.consume_user_activation();
+    }
+
+    // 7. Return promise, and run the remaining steps in parallel.
+    HTML::queue_global_task(HTML::Task::Source::UserInteraction, realm->global_object(), GC::create_function(heap(), [realm, err = error, pending_doc, element = GC::Ref<Element> { *this }, promise]() {
```
We've discussed this. You linked to a discussion about how "in parallell" steps should not even resolve promises even. You also said that you should probably change the crypto, so that it does queue global task, instead of deferred invoke.

theIDinside replied:
I suspect that it means the latter, which got me wondering - should there be a function that handles this, instead of having to queue a global task manually and then reject/resolve a promise, since this probably something that needs to happen in a *lot* of places?

AtkinsSJ replied:
"Queue a global task" is a specific spec algorithm. If the spec doesn't say to do it, we shouldn't do it.

theIDinside replied:
Well, I won't revert the changes, until you've OK'd it, since I'd be at risk to once again undo what was already undo once.

theIDinside commented on Libraries/LibWeb/DOM/Element.cpp:
```diff
@@ -1905,6 +1910,185 @@ WebIDL::ExceptionOr<void> Element::insert_adjacent_html(String const& position,
+    // 7. Return promise, and run the remaining steps in parallel.
+    HTML::queue_global_task(HTML::Task::Source::UserInteraction, realm->global_object(), GC::create_function(heap(), [realm, err = error, pending_doc, element = GC::Ref<Element> { *this }, promise]() {
+        HTML::TemporaryExecutionContext context(*realm, HTML::TemporaryExecutionContext::CallbacksEnabled::Yes);
+        RequestFullscreenError error = err;
```
You understood why I did it :P. I'll address it.

theIDinside commented on Libraries/LibWeb/DOM/Element.cpp:
```diff
@@ -3255,11 +3439,6 @@ size_t Element::ordinal_value() const
     return numbering;
 }
 
-bool Element::is_fullscreen_element() const
```
It is used. In the 1st commit. It's just changed in the later commits to do something else. I've taken the approach with these commits that they all must be buildable.

theIDinside replied:
Removed it from older commit and just returned false in `SelectorEngine` in the first commit instead.

theIDinside commented:
Fixed most of your nits @ADKaster. Rebased onto master.

teaalltr commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -6339,6 +6340,181 @@ bool Document::fullscreen_enabled() const
+        return;
+
+    // 2. Unfullscreen elements whose fullscreen flag is set, within document’s top layer, except for document’s fullscreen element.
+    Vector<GC::Ref<Element>, 8> fullscreen_elements;
```
`GC::MarkedVector<GC::Ref<Element>, 8>`  like this you mean @ADKaster ?

theIDinside commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -6357,6 +6359,187 @@ bool Document::fullscreen_enabled() const
+        iframe_element->set_iframe_fullscreen_flag(false);
+
+    remove_an_element_from_the_top_layer_immediately(element);
+    invalidate_layout_tree(InvalidateLayoutTreeReason::DocumentAddAnElementToTheTopLayer);
```
I guess. I guess we should rename the invalidation reason to catch both adding and removing? I felt it unnecessary to create a new value. So `InvalidateLayoutTreeReason::DocumentMutatedTheTopLayer` instead?

theIDinside replied:
Well, who cares I guess, I'll add a new `DocumentRemoveAnElementFromTheTopLayer`

AtkinsSJ commented on Libraries/LibWeb/DOM/Element.cpp:
```diff
@@ -2209,6 +2210,34 @@ GC::Ref<WebIDL::Promise> Element::request_fullscreen()
     return promise;
 }
 
+// https://fullscreen.spec.whatwg.org/#model
```
Yeah I've spent too long trying to directly link to spec sections. :laughing:

theIDinside commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -6357,6 +6359,187 @@ bool Document::fullscreen_enabled() const
+    bool resize = false;
+
+    // 4. Let docs be the result of collecting documents to unfullscreen given doc.
+    Vector<GC::Ptr<Document>> docs = doc->collect_documents_to_unfullscreen();
```
Technically I don't think neither needs to be, because neither runs any javascript at that point. The 2nd one definitely should be because `collect_documents_to_unfullscreen` actually returns a GC::RootVector.

The first one is just to stash the elements, because otherwise, I will invalidate the iterators when iterating over `top_layer_elements()`.

Do you still want me to change the first one to be a `GC::RootVector`?

gmta commented on Libraries/LibWebView/HeadlessWebView.h:
```diff
@@ -35,6 +35,9 @@ class HeadlessWebView : public WebView::ViewImplementation {
     Optional<String> m_pending_prompt_text;
 
+    // When restoring from fullscreen, we need to know to what dimension.
+    Web::DevicePixelRect m_cached_dimensions;
```
Calling this `m_cached_dimensions` seems to hint at some form of performance optimization. Maybe `m_previous_dimensions` is a better fit?

theIDinside replied:
A lot better, actually! Will change.

gmta commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -4454,6 +4454,9 @@ bool Document::is_allowed_to_use_feature(PolicyControlledFeature feature) const
         return true;
+    case PolicyControlledFeature::Fullscreen:
+        // FIXME: Implement the permissions policy specification
+        return true;
```
Commit message: why is the policy check "stupid"? If there's a good reason to call it that then add that reasoning to the commit message, if it's just subjective then leave it out please.

gmta commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -6331,6 +6332,88 @@ void Document::append_pending_fullscreen_change(PendingFullscreenEvent::Type typ
+    }
+
+    // 1. If this is a shadow root and its host is not connected, then return null.
+    // We're not a shadow root
```
This is not spec text, please prefix it with `// AD-HOC:`. But why is step 1 in the algo then?

ADKaster replied:
It's pretty unfortunate that DocumentOrShadowRoot doesn't correspond to a C++ "FooMixin" class like the other interface mixins we have do, but fixing that design problem seems out of scope of this change.

theIDinside replied:
Yeah, if/when we land this PR, I could gladly take a look at making that happen, if you want to, see if I can make some headway? I'll gladly be the "owner" of this going forward and make sure it stays up to date.

gmta commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -132,6 +132,7 @@
 #include <LibWeb/HTML/Scripting/Agent.h>
```
Typo in commit message: "Howevevr"

gmta commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -6414,6 +6417,196 @@ bool Document::fullscreen_enabled() const
+        if (!container)
+            break;
+        //  6. If container’s iframe fullscreen flag is set, break.
+        if (HTML::HTMLIFrameElement* iframe_element = as<HTML::HTMLIFrameElement>(container.ptr()); iframe_element) {
```
This `as<HTML::HTMLIFrameElement>` will panic if `container.ptr()` is not the right type, so you can forego the `iframe_element` condition and add the `iframe_fullscreen_flag()` condition instead.

theIDinside commented on Libraries/LibWeb/DOM/ShadowRoot.cpp:
```diff
@@ -31,6 +32,30 @@ void ShadowRoot::finalize()
+        return nullptr;
+
+    // 3. If candidate and this are in the same tree, then return candidate.
+    if (auto* retargeted_element = as<Element>(candidate); retargeted_element && &retargeted_element->root() == &root())
```
You mean that I'm being redundant here, because once we reach the if-statement, we can be guaranteed to have an `Element` type? I will remove the `retargeted_element &&` part.

theIDinside commented on Libraries/LibWeb/DOM/Document.cpp:
```diff
@@ -6331,6 +6332,88 @@ void Document::append_pending_fullscreen_change(PendingFullscreenEvent::Type typ
+    // element, null, and false.
+    auto hide_until = get_hide_until(showing_hint_popover_list());
+
+    // Address a bug in the spec https://github.com/whatwg/fullscreen/issues/245
```
I'll add the following comment

```
    // Finding topmost popover ancestor algorithm takes different parameters than those
    // described by the fullscreen spec. Since the new algorithm takes 4 parameters, with the new "popover list"
    // we must also account for the auto popover list.
    // More can be read about this "spec bug" in https://github.com/whatwg/fullscreen/issues/245
```
